### PR TITLE
Update cond_v2 design doc

### PR DESCRIPTION
**Pull request by skye**

### CODE REVIEW

**Comment by skye** [comment]
@saxenasaurabh

File: rfcs/20180507-cond-v2.md [hunk 1]
```diff
@@ -70,21 +75,17 @@ We don't want to lower If ops that will eventually be consumed by the [XLA encap
...
+This is the same method that tf.function and [while_v2](https://github.com/tensorflow/community/blob/master/rfcs/20180821-differentiable-functional-while.md) use for intermediate gradients, making all of them compose nicely.
 
-
-1.  Use placeholders for intermediates during construction, then use a C++ rewrite (Grappler or GraphOptimizationPass) to rewire the graph.
-1.  Output every possible intermediate.
-1.  It might already work as-is.
-    1.  Except for ExtendGraph -- solution could be to make C API and Session share Graph*
-
-**Update**: we went with (2) output every possible intermediate
+Alternatives considered:\
```
Comment (saxenasaurabh) [inline rfcs/20180507-cond-v2.md]: Nit: remove trailing backslash

File: rfcs/20180507-cond-v2.md [hunk 2]
```diff
@@ -93,26 +94,34 @@ After adding the intermediate outputs to the forward If op's branch functions, i
...
 
-Introduce a special op to output mirror tensors. This op's shape inference function will claim to output the same shape and type of the mirrored output, but since the tensor isn't actually needed the kernel will produce some small value to avoid producing large unnecessary values. If/when the op doesn't need to produce a value (e.g. via lowering + pruning), the kernel can CHECK or similar.
+In the branch where the intermediate is actually produced, the intermediate tensor is wrapped in an optional via the [`OptionalFromValue` op](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/ops/dataset_ops.cc#L629), meaning the output optional will contain the intermediate if that branch is taken. In the other branch, the "mirror" tensor is produced via the [`OptionalNone` op](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/ops/dataset_ops.cc#L635), meaning the output optional will have no value if the other branch is taken.
 
+Each gradient branch then only unwraps the optionals from its corresponding forward branch to pass to the gradient computation.
 
-### Taking the gradient of deserialized If ops
+Alternatives considered:\
```
Comment (saxenasaurabh) [inline rfcs/20180507-cond-v2.md]: Same as above

File: rfcs/20180507-cond-v2.md [hunk 3]
```diff
@@ -4,21 +4,26 @@
...
 
 ## Objective
 
 **Switch tf.cond to emit a single If op.**
 
-We can do tf.while_loop next.
-
-This would make mapping to XLA's control flow constructs easier/possible. In particular, just switching to the If op would be a big win (more work needed to get cond working with XLA than while_loop, which already had a lot of work done), and easier than while loop. It will also making debugging and analysis of cond constructs much simpler, e.g. to implement higher-order derivatives.
+Benefits:
+* Second-order derivatives
```
[the remark this answers is not in the document]
Comment (skye) [reply rfcs/20180507-cond-v2.md]: Done.

File: rfcs/20180507-cond-v2.md (hunk 1 above)
Comment (skye) [reply rfcs/20180507-cond-v2.md]: Done.

File: rfcs/20180507-cond-v2.md (hunk 2 above)
Comment (skye) [reply rfcs/20180507-cond-v2.md]: Done.

**Pull request merged**